CppImport/src/macro/LexicalHelper.cpp

```diff
@@ -0,0 +1,283 @@
+{
+	if (auto callExpr = clang::dyn_cast<clang::CallExpr>(clangAstNode))
+		correctNode(callExpr->getCallee()->getSourceRange(),
+											envisionAstNode);
+	else if (auto constrExpr = clang::dyn_cast<clang::CXXConstructExpr>(clangAstNode))
+	{
+		correctNode(
+				clang::SourceRange(constrExpr->getLocStart(),
+										 constrExpr->getParenOrBraceRange().getBegin()),
+											envisionAstNode);
+	}
+	else
+		correctNode(clangAstNode->getSourceRange(), envisionAstNode);
+}
+
+void LexicalHelper::correctNode(clang::SourceRange range, Model::Node* original)
+{
+	if (DCast<OOModel::ReturnStatement>(original)) return;
+	if (DCast<OOModel::NewExpression>(original)) return;
+	if (DCast<OOModel::BinaryOperation>(original)) return;
+	if (DCast<OOModel::IfStatement>(original)) return;
+
+	if (!clang_->isMacroRange(range)) return;
+
+	auto transformed = unexpandedSpelling(range);
+
+	if (DCast<OOModel::CastExpression>(original))
+	{
+		QRegularExpression regularExpression("^static_cast<\\s*((\\w|\\*|##)+)\\s*>");
+		auto match = regularExpression.match(transformed);
+
+		if (match.hasMatch())
+			transformed = match.captured(1);
+		else
+		{
+			qDebug() << "unmatched regex in cast expression" << transformed;
+			return;
+		}
+	}
+	else if (DCast<OOModel::MethodCallExpression>(original))
+	{
+		if (transformed.endsWith("(")) transformed = transformed.left(transformed.length() - 1);
+	}
+	else if (DCast<OOModel::VariableDeclaration>(original))
+	{
+		QRegularExpression regularExpression("^((\\w|##)+)");
+		auto match = regularExpression.match(transformed);
+
+		if (match.hasMatch())
+			transformed = match.captured(1);
+		else
+		{
+			qDebug() << "unmatched regex in vardecl" << transformed;
+			return;
+		}
+	}
+	else if (DCast<OOModel::Class>(original))
+	{
+		QRegularExpression regularExpression("^((\\w|##)+)");
+		auto match = regularExpression.match(transformed);
+
+		if (match.hasMatch())
+			transformed = match.captured(1);
+		else
+		{
+			qDebug() << "unmatched regex in class" << transformed;
+			return;
+		}
+	}
+
+	transformations_.insert(original, transformed);
+}
+
+bool LexicalHelper::contains(clang::SourceRange r, clang::SourceRange o)
+{
+	auto range = unexpandedSourceRange(r);
+	auto other = unexpandedSourceRange(o);
+
+	auto s = clang_->sourceManager()->getSpellingLoc(range.getBegin()).getPtrEncoding();
+	auto e = clang_->sourceManager()->getSpellingLoc(range.getEnd()).getPtrEncoding();
+	auto os = clang_->sourceManager()->getSpellingLoc(other.getBegin()).getPtrEncoding();
+
+	return s <= os && os <= e;
+}
+
+void LexicalHelper::applyLexicalTransformations(Model::Node* node, NodeMapping* mapping, QVector<QString> formalArgs)
+{
+	if (transformations_.contains(mapping->original(node)))
+	{
+		auto transformed = transformations_.value(mapping->original(node));
+
+		bool containsArg = false;
+		for (auto arg : formalArgs)
+			if (transformed.contains(arg))
+			{
+				containsArg = true;
+				break;
+			}
+
+		if (containsArg)
```

Comment:
Please explain what's going on here. Additionally explain what this method does in a method comment.